File: DonRyu/2022-10-18-Clone Graph.js
var cloneGraph = function(node) {
    // Nodes we have already copied
    const visited = {};

    // DFS function to copy graph
    const dfs = (node) => {
        if (!node) return node;
        // If we have seen this node before, return it
        if (visited[node.val]!=null) return visited[node.val];

        // Create base for copied node
        const root = new Node(node.val);
        // Add this copied node to group of nodes we hav copied
        visited[node.val] = root;

        // Add copied neighbors to the current copied node
        node.neighbors.forEach(n => root.neighbors.push(dfs(n)))
        return root;
    }

    // Return new copied graph
    return dfs(node);
};

File: BrianLee/Week6/2022-10-18-Clone-Graph.java
class Solution {
    public Node cloneGraph(Node node) {
        if(node == null) return null;
        Map<Integer,Node> nodes = new HashMap<>();
        return dfs(node, nodes);
    }

    private Node dfs(Node node, Map<Integer,Node> nodes) {
        if(nodes.containsKey(node.val)) return nodes.get(node.val);

        Node copy = new Node(node.val);
        nodes.put(node.val, copy);

        for(Node child : node.neighbors) {
            copy.neighbors.add(dfs(child, nodes));
        }

        return copy;
    }
}